cli2telegram/daemon.py

```diff
@@ -0,0 +1,132 @@
+LOGGER.setLevel(logging.DEBUG)
+handler = logging.StreamHandler(sys.stdout)
+handler.setLevel(logging.DEBUG)
+formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+handler.setFormatter(formatter)
+LOGGER.addHandler(handler)
+
+class Daemon:
+    def __init__(self, CONFIG:Config):
+        self.CONFIG = CONFIG
+        self.RUNNING = True
+        self.PIPE_NAME = CONFIG.DAEMON_PIPE_PATH.value
+        self.message_queue = queue.Queue()
+
+    def run(self):
+        try:
+            os.mkfifo(self.PIPE_NAME)
+        except OSError as ose:
+            LOGGER.debug(ose)
+            if os.path.exists(self.PIPE_NAME):
+                raise
+     
+        signal(SIGINT, self.signal_handler) # ctrl+c
+        signal(SIGTERM, self.signal_handler) # systemctl stop
+
+        threading.Thread(target=self.send_message_worker, daemon=True).start()
+
+        LOGGER.info("daemon listening on {}".format(self.PIPE_NAME))
+
+        while True:
+            with open(self.PIPE_NAME) as pipe:
+                LOGGER.debug("wating for input from pipe")
+                pipe_buffer = ''
+                while True:
+                    data = pipe.read()
+                    LOGGER.debug(f"received {data}")
+                    if len(data) == 0 and self.RUNNING:
+                        LOGGER.debug("No data from pipe. Sending buffer to telegram.");
+                        self.message_queue.put(pipe_buffer)
+                        pipe_buffer = ''
+                        break
+                    elif not self.RUNNING:
+                        break
+                    else:
+                        pipe_buffer += data
+            if not self.RUNNING:
+                break
+
+    def send_message_worker(self):
+        while True:
+            message = self.message_queue.get()
+            LOGGER.debug(f"Got {message} from queue. sending...")
+            self._try_send_message(message)
+            self.message_queue.task_done()
+
+    def _try_send_message(self, message: str):
+        """
+        Sends a message
+        :param message: the message to send
+        """
+        started_trying = datetime.now()
+        success = False
+        updater = Updater(token=self.CONFIG.TELEGRAM_BOT_TOKEN.value, use_context=True)
+        while not success:
+            try:
+                chat_id = self.CONFIG.TELEGRAM_CHAT_ID.value
+                send_message(updater.bot, chat_id, message, parse_mode="markdown")
+                success = True
+            except Exception as ex:
+                LOGGER.exception(ex)
+
+                if not self.CONFIG.RETRY_ENABLED.value:
+                    break
+
+                tried_for = datetime.now() - started_trying
+                if tried_for > self.CONFIG.RETRY_GIVE_UP_AFTER.value:
+                    LOGGER.warning(f"Giving up after trying for: {tried_for}. Placing message back in queue.")
+                    self.message_queue.put(message)
```

Comment:
Except for this small difference, this whole method is copied from the `cli.py`, right? I don't think the logic should be different for the daemon. If the message sending fails, just skip the queue item. If the user wants cli2telegram to retry endlessly they can just specify a very long "give up after" interval.

Also, we shouldn't copy 99% of the code for this one change. It should be possible to extract this method into the `util` package, where it can be used by both the cli and the daemon. The `CONFIG` variable is a singleton and can theoretically be used from anywhere, but I would pass the configuration values into the util function.

Reply:
Agreed. Will rework this.